Deprecate showcompact()

nalimilan opened the pull request:
There is no obvious reason to provide a special function for the `:compact` `IOContext` property
but not for other properties. `showcompact` used to be useful to print a single-line representation of an array at the REPL, but now `show` does the same thing. So it should only be needed for collections to print their elements when implementing `show`, and for tests (most of the changes here).

Also improve a few docstrings.

nalimilan committed:
Deprecate showcompact()

There is no obvious reason to provide a special function for the :compact IOContext property
but not for other properties. showcompact() used to be useful to print a single-line
representation of an array at the REPL, but now show() does the same thing. So it should only
be needed for collections to print their elements when implementing show(), and for tests
(most of the changes here).

Also improve a few docstrings.

helgee commented on NEWS.md:
```diff
@@ -1061,7 +1061,11 @@ Deprecated or removed
+
+  * `showcompact(io, x...)` has been deprecated in favor of
+    `show(IOContext(io, :compact => true), x...)` ([#26080]).
+    Use `sprint(showcompact, x...)` instead of `sprint(show, x..., context=:compact => true)`.
```
That's the wrong way around, isn't it?

nalimilan replied:
Thanks. I had originally written "is deprecated in favor of", but I changed it because it's not really a separate deprecation, mostly an advice. Will fix in the next round.

nalimilan replied:
Fixed.

JeffBezanson commented:
The downside is this has to allocate a new IOContext on every call, but I guess that's life.

vtjnash commented on base/strings/io.jl:
```diff
@@ -69,15 +69,21 @@ println(io::IO, xs...) = print(io, xs..., '\n')
 Call the given function with an I/O stream and the supplied extra arguments.
 Everything written to this I/O stream is returned as a string.
+`context` can be either an [`IOContext`](@ref) whose properties will be used,
+or a `Pair` specifying a property and its value. `sizehint` suggests the capacity
```
Specifying a Pair does happen to work, but only the usage case of passing an IO was what I had really originally intended (and found to be common).

I think that we maybe instead should suggest the usage of anonymous functions as a general way of passing a configuration set of arbitrarily many new properties:
```julia
showcompact = (io, v) -> show(IOContext(io, :compact => true), v)
sprint(showcompact, 66.66666)
```

nalimilan replied:
OK. I assumed it was intended since  the signature doesn't use `::IOContext`.

I actually started with the anonymous function approach, but it's quite verbose and much less convenient than passing a pair for common cases (have a look at the diff to get and idea). So I'd say we'd better continue to support passing a pair.

vtjnash replied:
I don't intend to prevent passing a Pair, but it simply doesn't generalize as well. Although I suppose it also quickly becomes easier just to write it out:
```julia
buf = IOBuffer(sizehint = sizehint)
io = IOContext(buf, :compact => true)
show(io, 66.66666)
return String(take!(buf))
```

nalimilan replied:
My thought was that passing a single pair is the most common case, and for other situations it's not too hard to create a full `IOContext`.

JeffBezanson replied:
I agree it's very nice to be able to pass a single pair.

ararslan commented:
If people are concerned about the `IOContext` part, perhaps we could add a `compact` keyword argument to `printstyled`?

JeffBezanson commented:
Triage is in favor of removing `showcompact`, and improving the documentation about how to implement it correctly via `show`.

nalimilan committed:
Add documentation for IOContext's :compact property in Custom pretty-printing section

nalimilan commented:
I've rebased and added a short presentation of `IOContext` and `:compact` in the "Custom pretty-printing" section of the manual. Let me know whether it's OK.

JeffBezanson commented:
LGTM!
